### PR TITLE
Add support for rendering more logical classsynopsis markup

### PR DESCRIPTION
**Pull request by Girgias**
while allowing the old markup to render.

@mallardduck @heiglandreas

I'll push an example of some doc-en changes that would render this markup.

From my testing (which is not the most extensive) nothing is broken.

### CODE REVIEW

**Comment by mallardduck** [comment]
So far the only "issue" i'm seeing is that this doesn't remove the redundant name from methods. AFAIK the current behaviour there is the desired result, one where local methods do not have the prefix of the class/interface they are within.

I think correcting this would require duplicating some of what is done for classes currently. So when the method is processed by `format_classsynopsis_methodsynopsis_methodname_text` it needs a reference of the current class like `$this->cchunk["classsynopsis"]["classname"]`. Doing a quick dump debug it looks like this is the default value of `false` instead of the current Interface name. Might be worth introducing a new `interfacename` field unless you just want to reuse the `classname`.

**Comment by Girgias** [comment]
> So far the only "issue" i'm seeing is that this doesn't remove the redundant name from methods. AFAIK the current behaviour there is the desired result, one where local methods do not have the prefix of the class/interface they are within.
> 
> I think correcting this would require duplicating some of what is done for classes currently. So when the method is processed by `format_classsynopsis_methodsynopsis_methodname_text` it needs a reference of the current class like `$this->cchunk["classsynopsis"]["classname"]`. Doing a quick dump debug it looks like this is the default value of `false` instead of the current Interface name. Might be worth introducing a new `interfacename` field unless you just want to reuse the `classname`.
> Screenshot 2023-07-06 at 12 49 59 PM

Yeah that was the issue, I reused the classname field as otherwise when you have an interface list it would potentially override stuff.

Also added support for ``<ooexception>`` while I was there and did some refactoring to group stuff together.

File: phpdotnet/phd/Package/PHP/XHTML.php [hunk 1]
```diff
@@ -7,13 +7,6 @@ abstract class Package_PHP_XHTML extends Package_Generic_XHTML {
         'appendix'              => 'format_container_chunk',
         'article'               => 'format_container_chunk',
         'book'                  => 'format_root_chunk',
-        'classname'             => array(
-            /* DEFAULT */          'span',
-            'ooclass'           => array(
-                /* DEFAULT */      'format_suppressed_tags',
-                'classsynopsisinfo' => 'format_classsynopsisinfo_ooclass_classname',
-            ),
-        ),
```
Comment (Girgias) [inline phpdotnet/phd/Package/PHP/XHTML.php:16]: For reference, this is useless, as it duplicated the same information from the generic XHTML format.

**Comment by mallardduck** [comment]
Think this is looking pretty good from testing on my end. Only "issue" is that (https://github.com/php/doc-en/pull/2563) is missing my previous changes for DS interfaces. So it reverts some items back to abstract modifiers I had removed from affected DS interface methods. Maybe not a big deal tho since rebase should fix that and I'll need to make adjustments so those work with this PR.

**Comment by Girgias** [comment]
@mallardduck merged this so feel free to fix the DS markup again. :)

**Comment by mallardduck** [comment]
@Girgias - does gen_stub need to still be updated and ran on things? If I follow correctly that should cover core PHP things but not extensions right? Currently those are hitting the assertion you put in here:
https://github.com/php/phd/pull/77/files#diff-a57f5cc81d692b04c0ca105933ad8bb4ea57f2d1d517f71160c847e15cc1753fR1009

Should I update those by hand too, or is that what gen_stub should cover?

**Comment by mallardduck** [comment]
Ok - so is it expected that I should see build errors currently then? I'm seeing:

```
php phd/render.php --docbook doc-base/.manual.xml --package PHP --format php
[14:15:25 - Rendering Style       ] Running full build
[14:15:25 - Indexing              ] Skipping indexing
[14:15:25 - Rendering Style       ] Running full build
[14:15:25 - Rendering Format      ] Starting PHP-Web rendering

Fatal error: Uncaught AssertionError: assert($this->cchunk['classsynopsis']['legacy'] === true) in /Users/danpock/GitProjects/the-php/phd/phpdotnet/phd/Package/Generic/XHTML.php:1009
Stack trace:
#0 /Users/danpock/GitProjects/the-php/phd/phpdotnet/phd/Package/Generic/XHTML.php(1009): assert(false, 'assert($this->c...')
#1 /Users/danpock/GitProjects/the-php/phd/phpdotnet/phd/Render.php(123): phpdotnet\phd\Package_Generic_XHTML->format_classsynopsisinfo(false, 'classsynopsisin...', Array, Array)
#2 /Users/danpock/GitProjects/the-php/phd/render.php(121): phpdotnet\phd\Render->execute(Object(phpdotnet\phd\Reader))
#3 {main}
  thrown in /Users/danpock/GitProjects/the-php/phd/phpdotnet/phd/Package/Generic/XHTML.php on line 1009
```

When I add a try-catch with good old dump-die, I see it's from `Traversable`:

```
array(23) {
  ["classsynopsis"]=>
  array(7) {
    ["close"]=>
    bool(false)
    ["classname"]=>
    string(11) "Traversable"
```

**Comment by Girgias** [comment]
AHHH right I had forgotten those got updated to use your rendering >_> yeah those need to be fixed... I suppose let's just wait for @kocsismate to finish updating gen_stub so that we can fix those interfaces too

**Comment by Lewiscowles1986** [comment]
I Just rebased https://github.com/php/doc-en/pull/2638 atop @kocsismate branch and it looks like the error still happens

```
Fatal error: Uncaught AssertionError: assert($this->cchunk['classsynopsis']['legacy'] === true) in /var/www/phd/phpdotnet/phd/Package/Generic/XHTML.php:1009
Stack trace:
#0 /var/www/phd/phpdotnet/phd/Package/Generic/XHTML.php(1009): assert(false, 'assert($this->c...')
#1 /var/www/phd/phpdotnet/phd/Render.php(125): phpdotnet\phd\Package_Generic_XHTML->format_classsynopsisinfo(false, 'classsynopsisin...', Array, Array)
#2 /var/www/phd/render.php(121): phpdotnet\phd\Render->execute(Object(phpdotnet\phd\Reader))
#3 {main}
  thrown in /var/www/phd/phpdotnet/phd/Package/Generic/XHTML.php on line 1009

```

Is there more work to do in #2611 or is there a way to make a branch from before this PR merged?

**Comment by Girgias** [comment]
> I Just rebased [php/doc-en#2638](https://github.com/php/doc-en/pull/2638) atop @kocsismate branch and it looks like the error still happens
> 
> ```
> Fatal error: Uncaught AssertionError: assert($this->cchunk['classsynopsis']['legacy'] === true) in /var/www/phd/phpdotnet/phd/Package/Generic/XHTML.php:1009
> Stack trace:
> #0 /var/www/phd/phpdotnet/phd/Package/Generic/XHTML.php(1009): assert(false, 'assert($this->c...')
> #1 /var/www/phd/phpdotnet/phd/Render.php(125): phpdotnet\phd\Package_Generic_XHTML->format_classsynopsisinfo(false, 'classsynopsisin...', Array, Array)
> #2 /var/www/phd/render.php(121): phpdotnet\phd\Render->execute(Object(phpdotnet\phd\Reader))
> #3 {main}
>   thrown in /var/www/phd/phpdotnet/phd/Package/Generic/XHTML.php on line 1009
> ```
> 
> Is there more work to do in #2611 or is there a way to make a branch from before this PR merged?

Are you sure you have the latest commits from doc-base, doc-en, and phd?

As I can render the docs without any issues.

Commands used from my phd clone:
```
php ../base/configure.php 
php render.php -d ../base/.manual.xml -P PHP -f xhtml -o ../output/
```

**Comment by Lewiscowles1986** [comment]
So I'm running from outside of the phd repo, but I am on `master` for `phd`, `doc-base` and `doc-en`.

This is also affecting another builder, who instead of highlighting was working around using a commit pin.

**Comment by Lewiscowles1986** [comment]
`doc-en` commit hash `184082814f268f881bf7834c5400e1981140d8ea` (master HEAD)
`doc-base` commit hash `15be9f9b25fe4b4b553bd7aed973966ba8103f4e` (master HEAD)
`phpd` commit hash 619c1c1a68dbc04c42e6b9216a61701a3b076a22` (master HEAD)

**Comment by Girgias** [comment]
Can you change the:
``assert($this->cchunk["classsynopsis"]["legacy"] === true);`` on line 1009 of  phd/Package/Generic/XHTML.php to
```php
if (!$this->cchunk["classsynopsis"]["legacy"] === true) { var_dump($name); }
```
This might give some info without needed to go through a debugger

**Comment by Girgias** [comment]
> @Girgias is this still waiting on @kocsismate to update gen_stub and rerun that? It seems a lot like what I reported last week.

It has been updated, and the interfaces have been updated. Just classes still use the "legacy" rendering for the most part, which is somewhat incompatible with the output as I want to let translation some time to catch up.

**Comment by Lewiscowles1986** [comment]
@kocsismate has not yet been merged, but I have done a 3 way branch including master, kocsismate and the new code, as well as tested this on just kocsismate and master. It would really help me to know which sha's are being used for the instructions above and if it works in a clean-install, or if there are maybe files or instructions I'm not following.

I Started this to troubleshoot if there was a nicer way to handle https://github.com/php/doc-en/pull/2638/files#diff-38801fa32c4bd7bd97aa3c87c138be36e126e0337b22aacf27ab57365357dc20R10 as I think being able to build from master, branch or tag (all of which I can clone shallow) would work.

The linked PR is not my work, but only works (same error as I've shown here) if pinning to a specific commit sha.

**Comment by mallardduck** [comment]
@Lewiscowles1986 - looks like the current build errors are coming from DS needing to be updated for the new format. I'll send the follow up PR to update those for @Girgias syntax improvements.

**Comment by mallardduck** [comment]
@Girgias - sent a PR: https://github.com/php/doc-en/pull/2641 to fix the DS docs.

@Lewiscowles1986 - just shorthand for Data Structures extension. https://www.php.net/manual/en/intro.ds.php

**Comment by Girgias** [comment]
ACK currently not in front of a computer but will do it ASAP

**Pull request merged**